### PR TITLE
[SPARK-32357][INFRA] Publish failed and succeeded test reports in GitHub Actions

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to report the failed and succeeded tests in GitHub Actions in order to improve the development velocity by leveraging [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report). See the example below:

Screen Shot 2020-08-13 at 8 17 52 PM

Note that we cannot just use [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report) in Apache Spark because PRs are from the forked repository, and GitHub secrets are unavailable for the security reason. This plugin and all similar plugins require to have the GitHub token that has the write access in order to post test results but it is unavailable in PRs.

To work around this limitation, I took this approach:

1. In workflow A, run the tests and upload the JUnit XML test results. GitHub provides to upload and download some files.
2. GitHub introduced new event type [`workflow_run`](https://github.blog/2020-08-03-github-actions-improvements-for-fork-and-pull-request-workflows/) 10 days ago. By leveraging this, it triggers another workflow B.
3. Workflow B is in the main repo instead of fork repo, and has the write access the plugin needs. In workflow B, it downloads the artifact uploaded from workflow A (from the forked repository).
4. Workflow B generates the test reports to port from JUnit xml files.
5. Workflow B looks up the PR and posts the test reports.

The `workflow_run` event is very new feature, and looks not so many GitHub Actions plugins support. In order to make this working with [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report), I had to fork two GitHub Actions plugins to use:
 - [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report) to have this custom fix: https://github.com/HyukjinKwon/action-surefire-report/commit/c96094cc35061fcf154a7cb46807f2f3e2339476
    It added `commit` argument to specify the commit to post the test reports. With `workflow_run`, it can access, in workflow B, to the commit from workflow A.

 - [dawidd6/action-download-artifact](https://github.com/dawidd6/action-download-artifact) to have this custom fix: https://github.com/HyukjinKwon/action-download-artifact/commit/750b71af351aba467757d7be6924199bb08db4ed
    It added the support of downloading all artifacts from workflow A, in workflow B. By default, it only supports to specify the name of artifact.

    Note that I was not able to use the official [actions/download-artifact](https://github.com/actions/download-artifact) because:
      - It does not support to download artifacts between different workflows, see also https://github.com/actions/download-artifact/issues/3. Once this issue is resolved, we can switch it back to [actions/download-artifact](https://github.com/actions/download-artifact).

I plan to make a pull request for both repositories so we don't have to rely on forks.

### Why are the changes needed?

Currently, it's difficult to check the failed tests. You should scroll down long logs from GitHub Actions logs.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Manually tested at: https://github.com/HyukjinKwon/spark/pull/17, https://github.com/HyukjinKwon/spark/pull/18, https://github.com/HyukjinKwon/spark/pull/19, https://github.com/HyukjinKwon/spark/pull/20, and master branch of my forked repository.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I am running other tests to make sure it includes PySpark test reports as well in this PR. I will add a comment here once all testing is finished. Currently, I only did a basic test as described in the PR description.

cc @viirya, @dongjoon-hyun, @gengliangwang, @dbtsai FYI

File: .github/workflows/master.yml
Comment (HyukjinKwon) [inline .github/workflows/master.yml:187]: This path is an exact copy from Jenkins configuration.

File: .github/workflows/master.yml
Comment (HyukjinKwon) [inline .github/workflows/master.yml:185]: I think I should file an infra ticket and handle this github token. I will proceed separately.

Comment (HyukjinKwon) [reply .github/workflows/master.yml:185]: It's in progress at INFRA-20628.

**Comment by gengliangwang** [comment]
@HyukjinKwon the report looks great. Thanks for the work!

File: .github/workflows/master.yml
Comment (gengliangwang) [inline .github/workflows/master.yml:183]: I saw your comment in https://github.com/apache/spark/pull/29169#issuecomment-667876128
Quick question here: Is there another working plugin?

Comment (HyukjinKwon) [reply .github/workflows/master.yml:183]: Nope.. this one was the only one given my few tries. I'm trying to figure out the github token issue. I'll bring up more info later.

File: .github/workflows/master.yml
Comment (viirya) [reply .github/workflows/master.yml]: What the issue here?

**Review by viirya — COMMENTED** [review]
The report looks nice.

**Comment by dongjoon-hyun** [comment]
It's great, @HyukjinKwon ! :)

**Comment by dbtsai** [comment]
Thanks @HyukjinKwon This will speed up our development velocity dramatically.

File: .github/workflows/master.yml
Comment (HyukjinKwon) [reply .github/workflows/master.yml]: I'll leave a comment below.

**Comment by HyukjinKwon** [comment]
Just to share the current status,

In [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report) plugin (and all other similar plugins), it leverages `GITHUB_TOKEN` that is set by default in GitHub Actions. It uses GitHub API to create [status checks](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-status-checks) via [here](https://github.com/ScaCap/action-surefire-report/blob/master/action.js#L42-L43) - it requires write permission to the repo. However, the permissions of `GITHUB_TOKEN` [does not cover the case when a PR was raised based on the fork](https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token).

There are many similar issues and questions, for example, in [codecov](https://github.com/codecov/codecov-action/issues/29) or [GitHub community](https://github.community/t/make-secrets-available-to-builds-of-forks/16166). In case of Codecov, they managed to remove the requirement of `GITHUB_TOKEN` at [here](https://github.com/codecov/codecov-action/issues/29#issuecomment-595062189). Basically they used existing GitHub Actions environment to verify in their service. This is not feasible in our case because the plugin is dependent of GitHub API to create the status checks directly.

I investigated this issue yesterday and concluded there's no clean workaround to make this working out of the box. 
I am currently investigating the feasibility of _potential_ alternatives. I am not sure if all of them work or not:

- Use one environment variable, for example, `TEST_REPORT_GITHUB_TOKEN` as a GitHub secret. And then, guide committers to set `TEST_REPORT_GITHUB_TOKEN` as a GitHub secret in their forks. Note that the contributors would be able to report the test results as their tokens don't have the write access to the repo.

- Just don't run this test report for now but wait until GitHub provides an alternative to work around this. There looks many requests such as [this](https://github.community/t/make-secrets-available-to-builds-of-forks/16166).

- Just generate a token that only has the permission to change the status checks, and hardcode it in the repo. At the worst case people abuse this token, the status of PRs are changed.

**Comment by HyukjinKwon** [comment]
BTW, let me know if you guys have a preference in one of the options. I can investigate fewer options in this case.

**Comment by HyukjinKwon** [comment]
adding @srowen, @gatorsmile, @cloud-fan as well to collect more feedback.

**Comment by srowen** [comment]
If there's not a clean way to do it, and this isn't essential but nice-to-have, then option 1 seems the most viable. It would be opt-in and only for people who want to take the time to set this up, so not sure it would cover much. But probably more useful than doing it only after the fact, or exposing a token.

**Comment by HyukjinKwon** [comment]
Sure, let me take a closer look for that approach. In worst case, we might have to drop this and go back to the original @viirya's approach at #29169.

**Comment by HyukjinKwon** [comment]
I am making some changes to demonstrate the problem in terms of the fork and PRs. Please ignore the changes made from now on. I will switch back from the draft later when this is ready for a review.

**Comment by HyukjinKwon** [comment]
The latest commit above (https://github.com/apache/spark/pull/29333/commits/2688f21f1852b3e6a577fa9292985b346b9bdf6d) contains the problem in terms of forked repos and PRs at [this workload](https://github.com/apache/spark/pull/29333/checks?check_run_id=947513630).
The current changes in this PR are exactly same as https://github.com/HyukjinKwon/spark/pull/16 at my own fork and PR in terms of how the [ScaCap/action-surefire-report](https://github.com/ScaCap/action-surefire-report) works.

**Comment by HyukjinKwon** [comment]
Okay, GitHub team suggested a scenario that it could work out of the box. It's a bit complicated then I thought but I will try. Yes, it is a nice-to-have but I think it is very worthwhile as it saves time of all Spark developers in their daily development and review at Apache Spark.

**Comment by HyukjinKwon** [comment]
I elaborated it in the PR description how it works and why I had to fork two plugins.

File: .github/workflows/master.yml
Comment (HyukjinKwon) [inline .github/workflows/master.yml:12]: This was already fixed.

File: .github/workflows/master.yml
Comment (HyukjinKwon) [inline .github/workflows/master.yml:29]: By doing this, it treats newlines as spaces, which  It makes the title prettier. Without this, [actions/upload-artifact](https://github.com/actions/upload-artifact) fails for unknown reason.

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/runs/980576471 should show the example of this PR. Before this PR is merged, the test results would not be reported because the workflow B (see PR description) should run in main repo, not in the PR.

**Comment by HyukjinKwon** [comment]
https://github.com/HyukjinKwon/spark/runs/980576471 should show the example of this PR. Once this example shows the results correctly, this PR should be good to go.

Before this PR is merged, the test results would not be reported because the workflow B (see PR description) should run in main repo, not in the PR.

**Comment by HyukjinKwon** [comment]
Okay the example [looks working fine](https://github.com/HyukjinKwon/spark/runs/980622718). This PR should be ready for a review and merged. cc @srowen, @gengliangwang, @dongjoon-hyun, @dbtsai, @viirya can you take a look please?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thanks, @HyukjinKwon .

**Comment by HyukjinKwon** [comment]
Thank you all !!

**Comment by HyukjinKwon** [comment]
I opened a PR to verify if this works well in the main commit (https://github.com/apache/spark/commit/5debde94019d46d4ab66f7927d9e5e8c4d16a7ec) and the PR (https://github.com/apache/spark/pull/29429).

File: .github/workflows/master.yml
Comment (viirya) [inline .github/workflows/master.yml:182]: If previous `Run tests` is passed without failure, do we still need run this? I remember Github Actions has some conditions other than `always()` can be used?

Comment (HyukjinKwon) [reply .github/workflows/master.yml:182]: Yeah, if the tests file, it should upload JUnit XML files and then report the failed test cases. GitHub Actions has things like `failure()` but I think we should run this always.

**Comment by HyukjinKwon** [comment]
Also, I would like to thank @cpintado from GitHub. He virtually guided me here a lot on this.

**Comment by gengliangwang** [comment]
A late LGTM. Thanks for the great work!

**Comment by maropu** [comment]
Thanks, @HyukjinKwon @cpintado ! Great work!

**Comment by HyukjinKwon** [comment]
Okay .. I checked that it works in both PR and commit. There look a bit of delay roughly 2 ~ 5 minutes to have the test report after the main tests are finished but I think this is good enough.

**Pull request closed without merging**
